skyrim/libs/benchmark/cds.py

```diff
@@ -0,0 +1,528 @@
+                (len(time), len(self.channels), len(self.CDS_LAT), len(self.CDS_LON))
+            ),
+            dims=["time", "channel", "lat", "lon"],
+            coords={
+                "time": time,
+                "channel": self.channels,
+                "lat": self.CDS_LAT,
+                "lon": self.CDS_LON,
+            },
+        )
+        for i, channel in enumerate(self.channels):
+            cds_id, cds_levtype, cds_level = CDS_Vocabulary.get(channel)
+            cache_path = self._download_cds_grib_to_cache(
+                time=time, variable=cds_id, levtype=cds_levtype, level=cds_level
+            )
+            da = xr.open_dataarray(
+                cache_path, engine="cfgrib", backend_kwargs={"indexpath": ""}
+            )
+            cds_dataarray[:, i, :, :] = da.values
+
+        return cds_dataarray
+
+    def _download_cds_grib_to_cache(
+        self,
+        time: list[datetime.datetime],
+        variable: str,
+        levtype: str,
+        level: str,
+    ) -> str:
+        """
+        Downloads a CDS GRIB file to the cache directory.
+        """
+
+        sha = hashlib.sha256(f"{variable}_{levtype}_{level}_{time}".encode())
+        filename = sha.hexdigest()
+        cache_path = os.path.join(self.cache, filename)
+
+        logger.debug(
+            f"Request: datetime: {time}, variable: {variable}, levtype: {levtype}"
+        )
+        logger.debug(f"Cache path: {cache_path}")
+
+        if not os.path.exists(cache_path):
+            request_body = self._build_request_body(
+                time=time, variable=variable, levtype=levtype, level=level
+            )
+            self.cds_client.retrieve(
+                name=levtype, request=request_body, target=cache_path
+            )
+            self.cached_files.append(cache_path)
+        else:
+            logger.debug(f"File already exists in cache: {cache_path}")
+
+        return cache_path
+
+    def _build_request_body(
+        self,
+        time: List[datetime.datetime],
+        variable: str,
+        levtype: str,
+        level: str,
+    ):
+        """Builds the request body for the CDS API using the provided year, month, day, and time."""
+
+        year = sorted(list(set(str(t.year) for t in time)))
+        month = sorted(list(set(f"{t.month:02d}" for t in time)))
+        day = sorted(list(set(f"{t.day:02d}" for t in time)))
+        time = sorted(list(set(t.strftime("%H:%M") for t in time)))
+
+        # build the request body
+        request_body = {
+            "variable": variable,
+            "product_type": "reanalysis",
+            "year": year,
+            "month": month,
+            "day": day,
+            "time": time,
+            "format": "grib",
+        }
+        if levtype == "reanalysis-era5-pressure-levels":
+            request_body["pressure_level"] = level
+        logger.debug(f"Request body: {request_body}")
+        return request_body
+
+    def _ymdt_to_datetime(
```

Comment:
if this is potentially usable for other things, agree. Also we can start using `arrow` for time range ops.

Reply:
maybe we can have a time utils module since we are dealing a lot with time manipulation